File: hw/zehuaZhang/hw2checklists/techbowA2checklistsObjC/techbowA2checklistsObjC/Checklist.m
//
//  Checklist.m
//  techbowA2checklistsObjC
//
//  Created by ZhangZehua on 11/22/15.
//  Copyright © 2015 ZhangZehua. All rights reserved.
//

#import "Checklist.h"

@implementation Checklist

- (instancetype)initWithName: (NSString *) name
{
    self = [self initWithName:name iconName:@"No Icon"];
    
    return self;
}

- (instancetype)initWithName:(NSString *)name iconName:(NSString *) iconName
{
    self = [super init];
    if (self) {
        _items = [[NSMutableArray<ChecklistItem *> alloc] init];
        _name = name;
        _iconName = iconName;
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    if (self) {
        _name = (NSString *)[coder decodeObjectForKey:@"Name"];
        _items = (NSMutableArray<ChecklistItem *> *)[[coder decodeObjectForKey:@"Items"]mutableCopy];
        _iconName = (NSString *)[coder decodeObjectForKey:@"IconName"];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)coder
{
    [coder encodeObject:_name forKey:@"Name"];
    [coder encodeObject:_items forKey:@"Items"];
    [coder encodeObject:_iconName forKey:@"IconName"];
}

- (NSInteger) countUncheckedItems {
    NSInteger count = 0;
    for (ChecklistItem* item in _items) {
        if (!item.checked) {
            count++;
        }
    }
    return count;
}

- (void) sortChecklistItems {
    
    _items = (NSMutableArray <ChecklistItem *> *)[[_items sortedArrayUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
        NSDate *date1 = [(ChecklistItem *) obj1 dueDate];
        NSDate *date2 = [(ChecklistItem *) obj2 dueDate];
        return [date1 compare: date2];
    }]mutableCopy];
}
@end

File: hw/zehuaZhang/hw2checklists/TechbbowA2CheckLists/Checklist.swift
//
//  Checklist.swift
//  TechbbowA2CheckLists
//
//  Created by ZhangZehua on 11/18/15.
//  Copyright © 2015 ZhangZehua. All rights reserved.
//

import UIKit

class Checklist: NSObject, NSCoding{
    var name = ""
    var items = [ChecklistItem]()
    var iconName: String
    
    convenience init(name: String) {
        self.init(name: name, iconName: "No Icon")
    }
    
    init(name: String, iconName: String) {
        self.name = name
        self.iconName = iconName
        super.init()
    }
    
    required init?(coder aDecoder: NSCoder) {
        name = aDecoder.decodeObjectForKey("Name") as! String
        items = aDecoder.decodeObjectForKey("Items") as! [ChecklistItem]
        iconName = aDecoder.decodeObjectForKey("IconName") as! String
        super.init()
    }
    
    func encodeWithCoder(aCoder: NSCoder) {
        aCoder.encodeObject(name, forKey: "Name")
        aCoder.encodeObject(items, forKey: "Items")
        aCoder.encodeObject(iconName, forKey: "IconName")
    }
    
    func countUncheckedItems() -> Int {
        var count = 0
        for item in items {
            if !item.checked {
                count++
            }
        }
        return count
    }
    
    func sortChecklistItems() {
        items.sortInPlace({
            item1, item2 in return
            item1.dueDate.compare(item2.dueDate) == NSComparisonResult.OrderedAscending
        })
    }
}
